### ISSUE TITLE
etcdserver: leader changed during cluster creation causing test failures

### ISSUE THREAD

**Issue by foot** [weaveworks/wksctl#237]
It intermittently causes CI failures in the `integration-tests-container` test[0]

Unclear if etcd leader should change during cluster creation too.

[0]
```
time="2020-06-22T20:25:34Z" level=info msg="running ../../../cmd/wksctl/wksctl [apply --cluster=/tmp/multimaster_test_866680275/cluster.yaml --machines=/tmp/multimaster_test_866680275/machines.yaml --config-directory=/tmp/multimaster_test_866680275 --verbose --controller-image=docker.io/weaveworks/wksctl-controller:v0.8.3-rc.4]"
time="2020-06-22T20:27:23Z" level=info msg="running ../../../cmd/wksctl/wksctl [kubeconfig --cluster=/tmp/multimaster_test_866680275/cluster.yaml --machines=/tmp/multimaster_test_866680275/machines.yaml]"
time="2020-06-22T20:27:23Z" level=info msg="running kubectl [get nodes -o json --kubeconfig=]"
time="2020-06-22T20:27:24Z" level=info msg="The cluster currently has 1 node(s)"
time="2020-06-22T20:27:54Z" level=info msg="running kubectl [get nodes -o json --kubeconfig=]"
time="2020-06-22T20:27:54Z" level=info msg="The cluster currently has 1 node(s)"
time="2020-06-22T20:28:24Z" level=info msg="running kubectl [get nodes -o json --kubeconfig=]"
time="2020-06-22T20:28:24Z" level=info msg="The cluster currently has 1 node(s)"
time="2020-06-22T20:28:54Z" level=info msg="running kubectl [get nodes -o json --kubeconfig=]"
time="2020-06-22T20:28:54Z" level=info msg="The cluster currently has 1 node(s)"
time="2020-06-22T20:29:24Z" level=info msg="running kubectl [get nodes -o json --kubeconfig=]"
--- FAIL: TestMultimasterSetup (291.36s)
    multimaster_test.go:408: 
        	Error Trace:	multimaster_test.go:408
        	            				multimaster_test.go:392
        	            				multimaster_test.go:322
        	Error:      	Received unexpected error:
        	            	exit status 1
        	Test:       	TestMultimasterSetup
        	Messages:   	Command /usr/local/bin/kubectl failed. STDOUT: {
        	            	    "apiVersion": "v1",
        	            	    "items": [],
        	            	    "kind": "List",
        	            	    "metadata": {
        	            	        "resourceVersion": "",
        	            	        "selfLink": ""
        	            	    }
        	            	}
        	            	
        	            	STDERR: Error from server: rpc error: code = Unavailable desc = etcdserver: leader changed
FAIL
FAIL	github.com/weaveworks/wksctl/test/integration/container	291.363s
make: *** [integration-tests-container] Error 1

Exited with code exit status 2
CircleCI received exit code 2
```

**Comment by bboreham** [comment]
I usually search for the word "flaky" or "flake", so here they are.

**Comment by bboreham** [comment]
Logs from etcd, on one run - it seems the very first leader hand-over is quick and uneventful, but then it gets bogged down and leadership switches a couple more times.

```
$ kubectl -n kube-system logs etcd-node0
2020-06-23 09:56:15.538160 I | etcdmain: etcd Version: 3.3.15
2020-06-23 09:56:15.538192 I | etcdmain: Git SHA: 94745a4ee
2020-06-23 09:56:15.538276 I | etcdmain: Go Version: go1.12.9
2020-06-23 09:56:15.538279 I | etcdmain: Go OS/Arch: linux/amd64
2020-06-23 09:56:15.538309 I | etcdmain: setting maximum number of CPUs to 2, total number of available CPUs is 2
2020-06-23 09:56:15.538387 I | embed: peerTLS: cert = /etc/kubernetes/pki/etcd/peer.crt, key = /etc/kubernetes/pki/etcd/peer.key, ca = , trusted-ca = /etc/kubernetes/pki/etcd/ca.crt, client-cert-auth = true, crl-file = 
2020-06-23 09:56:15.538976 I | embed: listening for peers on https://172.17.0.4:2380
2020-06-23 09:56:15.539027 I | embed: listening for client requests on 127.0.0.1:2379
2020-06-23 09:56:15.539042 I | embed: listening for client requests on 172.17.0.4:2379
2020-06-23 09:56:15.546388 I | etcdserver: name = node0
2020-06-23 09:56:15.546405 I | etcdserver: data dir = /var/lib/etcd
2020-06-23 09:56:15.546409 I | etcdserver: member dir = /var/lib/etcd/member
2020-06-23 09:56:15.546412 I | etcdserver: heartbeat = 100ms
2020-06-23 09:56:15.546414 I | etcdserver: election = 1000ms
2020-06-23 09:56:15.546416 I | etcdserver: snapshot count = 10000
2020-06-23 09:56:15.546425 I | etcdserver: advertise client URLs = https://172.17.0.4:2379
2020-06-23 09:56:15.556358 I | etcdserver: initial advertise peer URLs = https://172.17.0.4:2380
2020-06-23 09:56:15.565130 I | etcdserver: initial cluster = node0=https://172.17.0.4:2380
2020-06-23 09:56:15.607567 I | etcdserver: starting member 40fd14fa28910cab in cluster a6ea9ad1b116d02f
2020-06-23 09:56:15.616330 I | raft: 40fd14fa28910cab became follower at term 0
2020-06-23 09:56:15.623645 I | raft: newRaft 40fd14fa28910cab [peers: [], term: 0, commit: 0, applied: 0, lastindex: 0, lastterm: 0]
2020-06-23 09:56:15.623674 I | raft: 40fd14fa28910cab became follower at term 1
2020-06-23 09:56:15.797428 W | auth: simple token is not cryptographically signed
2020-06-23 09:56:15.916537 I | etcdserver: starting server... [version: 3.3.15, cluster version: to_be_decided]
2020-06-23 09:56:15.918150 I | embed: ClientTLS: cert = /etc/kubernetes/pki/etcd/server.crt, key = /etc/kubernetes/pki/etcd/server.key, ca = , trusted-ca = /etc/kubernetes/pki/etcd/ca.crt, client-cert-auth = true, crl-file = 
2020-06-23 09:56:16.082396 I | embed: listening for metrics on http://127.0.0.1:2381
2020-06-23 09:56:16.082761 I | etcdserver: 40fd14fa28910cab as single-node; fast-forwarding 9 ticks (election ticks 10)
2020-06-23 09:56:16.083228 I | etcdserver/membership: added member 40fd14fa28910cab [https://172.17.0.4:2380] to cluster a6ea9ad1b116d02f
2020-06-23 09:56:16.093857 I | raft: 40fd14fa28910cab is starting a new election at term 1
2020-06-23 09:56:16.093880 I | raft: 40fd14fa28910cab became candidate at term 2
2020-06-23 09:56:16.093888 I | raft: 40fd14fa28910cab received MsgVoteResp from 40fd14fa28910cab at term 2
2020-06-23 09:56:16.093895 I | raft: 40fd14fa28910cab became leader at term 2
2020-06-23 09:56:16.093899 I | raft: raft.node: 40fd14fa28910cab elected leader 40fd14fa28910cab at term 2
2020-06-23 09:56:16.157656 I | etcdserver: setting up the initial cluster version to 3.3
2020-06-23 09:56:16.166891 I | etcdserver: published {Name:node0 ClientURLs:[https://172.17.0.4:2379]} to cluster a6ea9ad1b116d02f
2020-06-23 09:56:16.167181 I | embed: ready to serve client requests
2020-06-23 09:56:16.168580 I | embed: serving client requests on 172.17.0.4:2379
2020-06-23 09:56:16.168836 I | embed: ready to serve client requests
2020-06-23 09:56:16.169971 I | embed: serving client requests on 127.0.0.1:2379
2020-06-23 09:56:16.186371 N | etcdserver/membership: set the initial cluster version to 3.3
2020-06-23 09:56:16.186512 I | etcdserver/api: enabled capabilities for version 3.3
2020-06-23 09:58:37.158933 I | etcdserver/membership: added member a41dd41eed6b1820 [https://172.17.0.5:2380] to cluster a6ea9ad1b116d02f
2020-06-23 09:58:37.158952 I | rafthttp: starting peer a41dd41eed6b1820...
2020-06-23 09:58:37.159480 I | rafthttp: started HTTP pipelining with peer a41dd41eed6b1820
2020-06-23 09:58:37.160399 I | rafthttp: started streaming with peer a41dd41eed6b1820 (writer)
2020-06-23 09:58:37.161476 I | rafthttp: started streaming with peer a41dd41eed6b1820 (writer)
2020-06-23 09:58:37.171988 I | rafthttp: started peer a41dd41eed6b1820
2020-06-23 09:58:37.172766 I | rafthttp: started streaming with peer a41dd41eed6b1820 (stream Message reader)
2020-06-23 09:58:37.173019 I | rafthttp: added peer a41dd41eed6b1820
2020-06-23 09:58:37.173857 I | rafthttp: started streaming with peer a41dd41eed6b1820 (stream MsgApp v2 reader)
2020-06-23 09:58:38.403861 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 267.393659ms, to a41dd41eed6b1820)
2020-06-23 09:58:38.404460 W | etcdserver: server is likely overloaded
2020-06-23 09:58:38.850502 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 1.802769ms, to a41dd41eed6b1820)
2020-06-23 09:58:38.850515 W | etcdserver: server is likely overloaded
2020-06-23 09:58:39.639893 W | raft: 40fd14fa28910cab stepped down to follower since quorum is not active
2020-06-23 09:58:39.639954 I | raft: 40fd14fa28910cab became follower at term 2
2020-06-23 09:58:39.639961 I | raft: raft.node: 40fd14fa28910cab lost leader 40fd14fa28910cab at term 2
2020-06-23 09:58:40.636772 I | raft: 40fd14fa28910cab is starting a new election at term 2
2020-06-23 09:58:40.636835 I | raft: 40fd14fa28910cab became candidate at term 3
2020-06-23 09:58:40.636859 I | raft: 40fd14fa28910cab received MsgVoteResp from 40fd14fa28910cab at term 3
2020-06-23 09:58:40.637046 I | raft: 40fd14fa28910cab [logterm: 2, index: 761] sent MsgVote request to a41dd41eed6b1820 at term 3
2020-06-23 09:58:41.634168 W | etcdserver: read-only range request "key:\"/registry/health\" " with result "error:context canceled" took too long (2.000566423s) to execute
WARNING: 2020/06/23 09:58:41 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2020-06-23 09:58:42.175584 W | rafthttp: health check for peer a41dd41eed6b1820 could not connect: dial tcp 172.17.0.5:2380: connect: connection refused (prober "ROUND_TRIPPER_SNAPSHOT")
2020-06-23 09:58:42.175621 W | rafthttp: health check for peer a41dd41eed6b1820 could not connect: dial tcp 172.17.0.5:2380: connect: connection refused (prober "ROUND_TRIPPER_RAFT_MESSAGE")
2020-06-23 09:58:42.344057 I | raft: 40fd14fa28910cab is starting a new election at term 3
2020-06-23 09:58:42.344078 I | raft: 40fd14fa28910cab became candidate at term 4
2020-06-23 09:58:42.344085 I | raft: 40fd14fa28910cab received MsgVoteResp from 40fd14fa28910cab at term 4
2020-06-23 09:58:42.344092 I | raft: 40fd14fa28910cab [logterm: 2, index: 761] sent MsgVote request to a41dd41eed6b1820 at term 4
2020-06-23 09:58:43.737517 I | raft: 40fd14fa28910cab is starting a new election at term 4
2020-06-23 09:58:43.737541 I | raft: 40fd14fa28910cab became candidate at term 5
2020-06-23 09:58:43.737549 I | raft: 40fd14fa28910cab received MsgVoteResp from 40fd14fa28910cab at term 5
2020-06-23 09:58:43.737555 I | raft: 40fd14fa28910cab [logterm: 2, index: 761] sent MsgVote request to a41dd41eed6b1820 at term 5
2020-06-23 09:58:44.787745 W | etcdserver: timed out waiting for read index response (local node might have slow network)
2020-06-23 09:58:44.787815 W | etcdserver: read-only range request "key:\"/registry/jobs/\" range_end:\"/registry/jobs0\" limit:500 " with result "error:etcdserver: request timed out" took too long (7.002675233s) to execute
2020-06-23 09:58:45.047749 I | rafthttp: peer a41dd41eed6b1820 became active
2020-06-23 09:58:45.047818 I | rafthttp: established a TCP streaming connection with peer a41dd41eed6b1820 (stream MsgApp v2 reader)
2020-06-23 09:58:45.053324 I | rafthttp: established a TCP streaming connection with peer a41dd41eed6b1820 (stream Message writer)
2020-06-23 09:58:45.059618 I | rafthttp: established a TCP streaming connection with peer a41dd41eed6b1820 (stream MsgApp v2 writer)
2020-06-23 09:58:45.059980 I | rafthttp: established a TCP streaming connection with peer a41dd41eed6b1820 (stream Message reader)
2020-06-23 09:58:45.436497 I | raft: 40fd14fa28910cab is starting a new election at term 5
2020-06-23 09:58:45.436515 I | raft: 40fd14fa28910cab became candidate at term 6
2020-06-23 09:58:45.436521 I | raft: 40fd14fa28910cab received MsgVoteResp from 40fd14fa28910cab at term 6
2020-06-23 09:58:45.436526 I | raft: 40fd14fa28910cab [logterm: 2, index: 761] sent MsgVote request to a41dd41eed6b1820 at term 6
2020-06-23 09:58:45.442869 I | raft: 40fd14fa28910cab received MsgVoteResp from a41dd41eed6b1820 at term 6
2020-06-23 09:58:45.442891 I | raft: 40fd14fa28910cab [quorum:2] has received 2 MsgVoteResp votes and 0 vote rejections
2020-06-23 09:58:45.442902 I | raft: 40fd14fa28910cab became leader at term 6
2020-06-23 09:58:45.442908 I | raft: raft.node: 40fd14fa28910cab elected leader 40fd14fa28910cab at term 6
2020-06-23 09:58:45.443801 W | etcdserver: read-only range request "key:\"/registry/pods/kube-system/kube-scheduler-node1\" " with result "error:etcdserver: leader changed" took too long (1.248165759s) to execute
2020-06-23 09:58:45.444009 W | etcdserver: read-only range request "key:\"/registry/namespaces/default\" " with result "error:etcdserver: leader changed" took too long (3.757273021s) to execute
2020-06-23 09:58:45.444121 W | etcdserver: read-only range request "key:\"/registry/persistentvolumes\" range_end:\"/registry/persistentvolumet\" count_only:true " with result "error:etcdserver: leader changed" took too long (4.12191441s) to execute
2020-06-23 09:58:45.444304 W | etcdserver: read-only range request "key:\"/registry/minions/\" range_end:\"/registry/minions0\" " with result "error:etcdserver: leader changed" took too long (6.155513679s) to execute
2020-06-23 09:58:45.444497 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kube-system/kube-controller-manager\" " with result "error:etcdserver: leader changed" took too long (6.260337816s) to execute
2020-06-23 09:58:45.444938 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kube-system/kube-scheduler\" " with result "error:etcdserver: leader changed" took too long (6.567973021s) to execute
2020-06-23 09:58:45.444965 W | etcdserver: read-only range request "key:\"/registry/minions/\" range_end:\"/registry/minions0\" limit:500 " with result "error:etcdserver: leader changed" took too long (6.79602624s) to execute
2020-06-23 09:58:45.445252 W | etcdserver: read-only range request "key:\"/registry/minions/node1\" " with result "error:etcdserver: leader changed" took too long (6.463289406s) to execute
2020-06-23 09:58:45.447260 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kube-system/kube-controller-manager\" " with result "error:etcdserver: leader changed" took too long (6.470383358s) to execute
2020-06-23 09:58:45.447554 W | etcdserver: read-only range request "key:\"/registry/configmaps/kube-system/extension-apiserver-authentication\" " with result "error:etcdserver: leader changed" took too long (6.563988868s) to execute
2020-06-23 09:58:46.519513 W | etcdserver: read-only range request "key:\"/registry/pods/kube-system/kube-apiserver-node1\" " with result "error:context canceled" took too long (974.821341ms) to execute
WARNING: 2020/06/23 09:58:46 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2020-06-23 09:58:46.519727 W | etcdserver: read-only range request "key:\"/registry/pods/kube-system/kube-proxy-pwhjm\" " with result "error:context canceled" took too long (1.068188876s) to execute
WARNING: 2020/06/23 09:58:46 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2020-06-23 09:58:46.519768 W | etcdserver: read-only range request "key:\"/registry/pods/kube-system/etcd-node1\" " with result "error:context canceled" took too long (329.456943ms) to execute
WARNING: 2020/06/23 09:58:46 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2020-06-23 09:58:48.911744 W | etcdserver: read-only range request "key:\"/registry/pods/kube-system/weave-net-mn4th\" " with result "error:context canceled" took too long (3.468143177s) to execute
WARNING: 2020/06/23 09:58:48 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2020-06-23 09:58:48.911831 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kube-system/kube-scheduler\" " with result "error:context canceled" took too long (1.439460025s) to execute
WARNING: 2020/06/23 09:58:48 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2020-06-23 09:58:49.329037 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 42.988514ms, to a41dd41eed6b1820)
2020-06-23 09:58:49.329130 W | etcdserver: server is likely overloaded
2020-06-23 09:58:49.391918 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kube-system/kube-controller-manager\" " with result "error:context canceled" took too long (1.919505389s) to execute
WARNING: 2020/06/23 09:58:49 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2020-06-23 09:58:51.360286 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 108.965443ms, to a41dd41eed6b1820)
2020-06-23 09:58:51.435228 W | etcdserver: server is likely overloaded
2020-06-23 09:58:52.267111 W | etcdserver: read-only range request "key:\"/registry/health\" " with result "error:context canceled" took too long (2.629937439s) to execute
WARNING: 2020/06/23 09:58:52 grpc: Server.processUnaryRPC failed to write status: connection error: desc = "transport is closing"
2020-06-23 09:58:52.576186 W | etcdserver: failed to send out heartbeat on time (exceeded the 100ms timeout for 899.474778ms, to a41dd41eed6b1820)
2020-06-23 09:58:52.576199 W | etcdserver: server is likely overloaded
2020-06-23 09:58:52.785379 W | etcdserver: timed out waiting for read index response (local node might have slow network)
2020-06-23 09:58:52.786086 W | etcdserver: read-only range request "key:\"/registry/validatingwebhookconfigurations\" range_end:\"/registry/validatingwebhookconfigurationt\" count_only:true " with result "error:etcdserver: request timed out" took too long (7.700841992s) to execute
2020-06-23 09:58:52.789937 W | etcdserver: read-only range request "key:\"/registry/namespaces/default\" " with result "range_response_count:1 size:172" took too long (7.28232403s) to execute
2020-06-23 09:58:52.792067 W | etcdserver: read-only range request "key:\"/registry/cluster.k8s.io/machinedeployments\" range_end:\"/registry/cluster.k8s.io/machinedeploymentt\" count_only:true " with result "range_response_count:0 size:5" took too long (218.320162ms) to execute
2020-06-23 09:58:52.792332 W | etcdserver: read-only range request "key:\"/registry/networkpolicies\" range_end:\"/registry/networkpoliciet\" count_only:true " with result "range_response_count:0 size:5" took too long (219.148547ms) to execute
2020-06-23 09:58:52.792484 W | etcdserver: read-only range request "key:\"/registry/storageclasses\" range_end:\"/registry/storageclasset\" count_only:true " with result "range_response_count:0 size:5" took too long (219.307358ms) to execute
2020-06-23 09:58:52.792639 W | etcdserver: read-only range request "key:\"/registry/events/kube-system/kube-scheduler-node0.161b23ba0d72f820\" " with result "range_response_count:1 size:500" took too long (2.351621738s) to execute
2020-06-23 09:58:52.792826 W | etcdserver: read-only range request "key:\"/registry/pods/kube-system/kube-scheduler-node0\" " with result "range_response_count:1 size:1629" took too long (2.352039048s) to execute
2020-06-23 09:58:52.792956 W | etcdserver: read-only range request "key:\"/registry/clusterroles\" range_end:\"/registry/clusterrolet\" count_only:true " with result "range_response_count:0 size:7" took too long (2.707849941s) to execute
2020-06-23 09:58:52.793108 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kube-system/kube-controller-manager\" " with result "range_response_count:1 size:445" took too long (3.984128011s) to execute
2020-06-23 09:58:52.793286 W | etcdserver: read-only range request "key:\"/registry/serviceaccounts\" range_end:\"/registry/serviceaccountt\" count_only:true " with result "range_response_count:0 size:7" took too long (5.307616319s) to execute
2020-06-23 09:58:52.793433 W | etcdserver: read-only range request "key:\"/registry/podtemplates\" range_end:\"/registry/podtemplatet\" count_only:true " with result "range_response_count:0 size:5" took too long (5.320882317s) to execute
2020-06-23 09:58:52.793584 W | etcdserver: read-only range request "key:\"/registry/apiregistration.k8s.io/apiservices\" range_end:\"/registry/apiregistration.k8s.io/apiservicet\" count_only:true " with result "range_response_count:0 size:7" took too long (5.435975899s) to execute
2020-06-23 09:58:52.793738 W | etcdserver: read-only range request "key:\"/registry/minions/node1\" " with result "range_response_count:1 size:2617" took too long (6.097666424s) to execute
2020-06-23 09:58:52.793932 W | etcdserver: read-only range request "key:\"/registry/minions/node1\" " with result "range_response_count:1 size:2617" took too long (6.097873866s) to execute
2020-06-23 09:58:52.794092 W | etcdserver: read-only range request "key:\"/registry/pods/kube-system/etcd-node1\" " with result "range_response_count:1 size:1751" took too long (6.231313502s) to execute
2020-06-23 09:58:52.794236 W | etcdserver: read-only range request "key:\"/registry/services/endpoints/kube-system/kube-scheduler\" " with result "range_response_count:1 size:427" took too long (7.195043682s) to execute
```

**Comment by luxas** [comment]
What k8s release are you using? v1.14 or v1.16?
See https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/constants/constants.go#L429-L438 for a list of officially supported etcd releases. I'd upgrade etcd release if possible (IDK if it's set anywhere in wksctl atm)

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go#L134-L145 contains a relevant note for this too.

**Comment by luxas** [comment]
From https://github.com/etcd-io/etcd/blob/master/CHANGELOG-3.3.md

> The minimum recommended etcd versions to run in production are 3.2.28+, 3.3.18+, and 3.4.2+.

Latest 3.3.x is v3.3.22

**Comment by bboreham** [comment]
It's Kubernetes 1.16.11 since everything before that got broken on Wednesday.
Etcd is 3.3.15.

**Comment by bboreham** [comment]
The version of etcd seems to be coming from kubeadm? 
The version we pull down to local repository changed here, but we don't set that version in any Kubernetes parameters: https://github.com/weaveworks/wksctl/pull/232/files

**Comment by bboreham** [comment]
I merged a change in #175 to ignore such errors from kubectl.
